File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.java
class Solution {
    int[] dr = {0, 0, 1, -1};
    int[] dc = {1, -1, 0, 0};

    public int longestIncreasingPath(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        int[][] memo = new int[m][n];
        int result = 0;

        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                result = Math.max(result, dfs(r, c, matrix, memo));
            }
        }

        return result;
    }

    private int dfs(int row, int col, int[][] matrix, int[][] memo) {
        if (memo[row][col] != 0) return memo[row][col];

        int max = 0;
        for (int i = 0; i < 4; i++) {
            int nr = row + dr[i];
            int nc = col + dc[i];

            if (nr < 0 ||
                nr >= matrix.length ||
                nc < 0 ||
                nc >= matrix[0].length ||
                matrix[row][col] >= matrix[nr][nc]
            ) continue;

            max = Math.max(max, dfs(nr, nc, matrix, memo));
        }

        memo[row][col] = 1 + max;

        return memo[row][col];
    }
}

File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.kt
class Solution {
    private val dr = intArrayOf(0, 0, 1, -1)
    private val dc = intArrayOf(1, -1, 0, 0)

    fun longestIncreasingPath(matrix: Array<IntArray>): Int {
        val m = matrix.size
        val n = matrix[0].size
        val memo = Array(m) { IntArray(n) }

        return (0 until m).maxOf { r ->
            (0 until n).maxOf { c ->
                dfs(r, c, matrix, memo)
            }
        }
    }

    private fun dfs(row: Int, col: Int, matrix: Array<IntArray>, memo: Array<IntArray>): Int {
        if (memo[row][col] != 0) return memo[row][col]

        var max = 0
        for (i in 0 until 4) {
            val nr = row + dr[i]
            val nc = col + dc[i]

            if (nr < 0 ||
                nr >= matrix.size ||
                nc < 0 ||
                nc >= matrix[0].size ||
                matrix[row][col] >= matrix[nr][nc]
            ) continue

            max = maxOf(max, dfs(nr, nc, matrix, memo))
        }

        memo[row][col] = 1 + max

        return memo[row][col]
    }
}